File: src/main/java/com/baidu/algorithm/binarysearch/_154_Find_Minimum_in_Rotated_Sorted_Array_II.py
class Solution(object):
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0

        begin, end = 0, len(nums) - 1
        while begin < end:
            mid = (begin + end) >> 1
            if nums[mid] == nums[end]:
                end -= 1
            elif nums[mid] > nums[end]:
                begin = mid + 1
            else:
                end = mid

        return nums[begin]

File: src/main/java/com/baidu/algorithm/binarysearch/_154_Find_Minimum_in_Rotated_Sorted_Array_II.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.binarysearch;

/**
 * _154_Find_Minimum_in_Rotated_Sorted_Array_II
 *
 * @author xuhaoran01
 */
public class _154_Find_Minimum_in_Rotated_Sorted_Array_II {

    public int findMin(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }

        int left = 0, right = nums.length - 1;
        while (left < right) {
            int mid = (left + right) >> 1;
            if (nums[mid] == nums[right]) {
                right--;
            } else if (nums[mid] > nums[right]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        return nums[left];
    }
}
